modules/schema/src/main/java/org/apache/ignite/internal/schema/marshaller/schema/AbstractSchemaSerializer.java

```diff
@@ -0,0 +1,134 @@
+    /** Schema serializer version. */
+    protected final short version;
+
+    /** Previous version serializer. */
+    protected final AbstractSchemaSerializer previous;
+
+    /**
+     * @param ver Serializer version.
+     * @param previous Previous version serializer.
+     */
+    protected AbstractSchemaSerializer(short ver, AbstractSchemaSerializer previous) {
+        this.version = ver;
+        this.previous = previous;
+    }
+
+    /**
+     * @param ver Serializer version.
+     */
+    protected AbstractSchemaSerializer(short ver) {
+        this(ver, null);
+    }
+
+    /**
+     * @return Serializer version;
+     */
+    public short getVersion() {
+        return version;
+    }
+
+    /**
+     * Serialize SchemaDescriptor object to byte array.
+     *
+     * @param desc SchemaDescriptor object.
+     * @return SchemaDescriptor byte array representation.
+     */
+    public byte[] serialize(SchemaDescriptor desc) {
+        return this.bytes(desc, createByteBuffer(desc)).array();
+    }
+
+    /**
+     * Deserialize byte array to SchemaDescriptor object.
+     *
+     * @param bytes SchemaDescriptor byte array representation.
+     * @return SchemaDescriptor object.
+     */
+    public SchemaDescriptor deserialize(byte[] bytes) {
+        ExtendedByteBuffer buf = createByteBuffer(bytes);
+
+        short ver = readVersion(buf);
+
+        return getSerializerByVersion(ver).value(buf);
+    }
+
+    /**
+     * Gets schema serializer by version.
+     *
+     * @param ver SchemaSerializer target version.
+     * @return SchemaSerializer object.
+     * @throws IllegalArgumentException If SchemaSerializer with right version is not found.
+     */
+    private SchemaSerializer getSerializerByVersion(short ver) {
+        if (ver == this.version)
+            return this;
+        else if (this.previous == null)
+            throw new IllegalArgumentException("Unable to find schema serializer with version " + ver);
+
+        return this.previous.getSerializerByVersion(ver);
+    }
+
+    /**
+     * Calculate size in bytes of SchemaDescriptor object.
+     *
+     * @param desc SchemaDescriptor object.
+     * @return Size in bytes.
+     */
+    private int size(SchemaDescriptor desc) {
+        return bytes(desc, ExtendedByteBufferFactory.calcSize()).size();
+    }
+
+    /**
+     * Wraps a byte array into a buffer.
+     *
+     * @param bytes Byte array.
+     * @return ByteBuffer byte buffer.
+     */
+    protected ExtendedByteBuffer createByteBuffer(byte[] bytes) {
+        return ExtendedByteBufferFactory.wrap(bytes);
+    }
+
+    /**
+     * Calculates the required size from SchemaDescriptor object and allocates a new ByteBuffer with the given size.
+     *
+     * @param desc SchemaDescriptor object.
+     * @return ByteBuffer object.
+     */
+    protected ExtendedByteBuffer createByteBuffer(SchemaDescriptor desc) {
+        return ExtendedByteBufferFactory.allocate(size(desc));
```

Comment:
Looks weird.
You create a fake buffer just to calculate size, and then allocate expandable buffer for serialization.
Why don't preallocate some buffer? or have a method for size calculation.

Reply:
Now I have method 'size' for size calculation, thanks.